ddl/schema.go

```diff
@@ -70,6 +70,49 @@ func onCreateSchema(t *meta.Meta, job *model.Job) (ver int64, _ error) {
 	}
 }
 
+func rollbackDropSchema(t *meta.Meta, job *model.Job, dbInfo *model.DBInfo) (ver int64, err error) {
+	// If dbInfo schema state is public, no need to rollback.
+	if dbInfo.State == model.StatePublic {
+		job.State = model.JobStateRollbackDone
+		job.SchemaState = dbInfo.State
+		return ver, nil
+	}
+	// Recover database.
+	// Change type to ActionCreateSchema for infoschema to load schema diff.
+	job.Type = model.ActionCreateSchema
+	defer func() {
+		job.Type = model.ActionDropSchema
+	}()
+	dbInfo.State = model.StatePublic
+	err = t.UpdateDatabase(dbInfo)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+	ver, err = updateSchemaVersion(t, job)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+
+	// Recover table.
+	var tables []*model.TableInfo
+	tables, err = t.ListTables(job.SchemaID)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+	// Change type to ActionCreateTable for infoschema to load schema diff.
+	job.Type = model.ActionCreateTable
+	for _, tblInfo := range tables {
+		job.TableID = tblInfo.ID
+		ver, err = updateSchemaVersion(t, job)
+		if err != nil {
+			return ver, errors.Trace(err)
+		}
+	}
+	job.State = model.JobStateRollbackDone
+	job.SchemaState = dbInfo.State
```

Comment:
Also need to add here `job.FinishTableJob(model.JobStateRollbackDone, model.StatePublic, ver, tblInfo)`?

Reply:
Great catch. I will fix it right now.